send-transaction-service/src/transaction_client.rs

```diff
@@ -115,13 +140,143 @@ where
         let leader_info = leader_info_provider.get_leader_info();
-        let leader_addresses = self.get_tpu_addresses(leader_info);
+        let leader_addresses = self.get_unique_tpu_addresses(leader_info);
         addresses.extend(leader_addresses);
 
         for address in &addresses {
             self.send_transactions(address, wire_transactions.clone(), stats);
         }
     }
+
+    fn protocol(&self) -> Protocol {
+        self.connection_cache.protocol()
+    }
+}
+
+impl<T> Cancelable for ConnectionCacheClient<T>
+where
+    T: TpuInfoWithSendStatic,
+{
+    fn cancel(&self) {}
+}
+
+pub struct SendTransactionServiceLeaderUpdater<T: TpuInfoWithSendStatic> {
+    leader_info_provider: CurrentLeaderInfo<T>,
+    my_tpu_address: SocketAddr,
+    tpu_peers: Option<Vec<SocketAddr>>,
+}
+
+#[async_trait]
+impl<T> LeaderUpdater for SendTransactionServiceLeaderUpdater<T>
+where
+    T: TpuInfoWithSendStatic,
+{
+    fn next_leaders(&mut self, lookahead_leaders: usize) -> Vec<SocketAddr> {
+        let discovered_peers = self
+            .leader_info_provider
+            .get_leader_info()
+            .map(|leader_info| {
+                leader_info.get_leader_tpus(lookahead_leaders as u64, Protocol::QUIC)
+            })
+            .filter(|addresses| !addresses.is_empty())
+            .unwrap_or_else(|| vec![&self.my_tpu_address]);
+        let mut all_peers = self.tpu_peers.clone().unwrap_or_default();
+        all_peers.extend(discovered_peers.into_iter().cloned());
+        all_peers
+    }
+    async fn stop(&mut self) {}
+}
+
+#[derive(Clone)]
+pub struct TpuClientNextClient {
+    runtime_handle: Handle,
+    sender: Sender<TransactionBatch>,
+    cancel: CancellationToken,
+}
+
+impl Cancelable for TpuClientNextClient {
+    fn cancel(&self) {
+        self.cancel.cancel();
+    }
+}
+
+impl TransactionClient for TpuClientNextClient {
+    fn send_transactions_in_batch(
+        &self,
+        wire_transactions: Vec<Vec<u8>>,
+        stats: &SendTransactionServiceStats,
+    ) {
+        let mut measure = Measure::start("send-us");
+        self.runtime_handle.spawn({
+            let sender = self.sender.clone();
+            async move {
+                let res = sender.send(TransactionBatch::new(wire_transactions)).await;
+                if res.is_err() {
+                    warn!("Failed to send transaction to channel: it is closed.");
+                }
+            }
+        });
+
+        measure.stop();
+        stats.send_us.fetch_add(measure.as_us(), Ordering::Relaxed);
+        stats.send_attempt_count.fetch_add(1, Ordering::Relaxed);
+    }
+
+    fn protocol(&self) -> Protocol {
+        Protocol::QUIC
+    }
+}
+
+pub fn spawn_tpu_client_send_txs<T>(
+    runtime_handle: Handle,
+    my_tpu_address: SocketAddr,
+    tpu_peers: Option<Vec<SocketAddr>>,
+    leader_info: Option<T>,
+    leader_forward_count: u64,
+    validator_identity: Option<Keypair>,
```

Comment:
This will be replaced with `client_certificate: QuicClientCertificate,` in the follow up commit (requires changes on the tpu-client-next side from  https://github.com/anza-xyz/agave/pull/3576/commits/2d146d4c14d2df438dcba67fc2f4675770a7c0a4